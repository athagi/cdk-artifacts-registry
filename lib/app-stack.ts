import * as cdk from '@aws-cdk/core';
import ec2 = require('@aws-cdk/aws-ec2');
// import * as ecr from '@aws-cdk/aws-ecr';

// class AppStack extends cdk.Stack {
export class AppStack extends cdk.Stack {
  public readonly vpc: ec2.Vpc;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // this.vpc = new ec2.Vpc(this, 'Vpc', { maxAzs: 1 });
  }
}
