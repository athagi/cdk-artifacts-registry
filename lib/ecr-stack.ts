import * as ecr from "@aws-cdk/aws-ecr";
import * as cdk from "@aws-cdk/core";

// ecr stack
export interface RepostiroiesStacksProps extends cdk.StackProps {
  repoNames: string[];
  lifecycleRule: ecr.LifecycleRule;
  prefix?: string;
}

export class RepositoriesStack extends cdk.Stack {
  public readonly repository: ecr.Repository;
  static LIFECYCLERULE: ecr.LifecycleRule = {
    maxImageCount: 10,
    rulePriority: 1,
    tagStatus: ecr.TagStatus.ANY,
  };

  constructor(
    scope: cdk.Construct,
    id: string,
    props: RepostiroiesStacksProps
  ) {
    super(scope, id, props);

    const repositories: string[] = props.repoNames;
    let prefix = "";
    if (
      props.prefix &&
      props.prefix.length > 0 &&
      !props.prefix.match(/(master|prod|production|main)/)
    ) {
      prefix = `${props.prefix}-`;
    }
    repositories.forEach((repo) => {
      this.createRepository(`${prefix}${repo}`, props.lifecycleRule);
    });
  }

  createRepository(
    repoName: string,
    lifecycleRule: ecr.LifecycleRule
  ): ecr.Repository {
    const repository = new ecr.Repository(this, repoName, {
      imageScanOnPush: true,
      repositoryName: repoName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    repository.addLifecycleRule(lifecycleRule);
    return repository;
  }
}
