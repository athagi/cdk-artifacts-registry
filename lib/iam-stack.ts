import * as iam from "@aws-cdk/aws-iam";
import * as cdk from "@aws-cdk/core";
export interface IamUserStacksProps extends cdk.StackProps {
  userNames: string[];
  strictedCidrs: string[];
  groupName: string;
  prefix?: string;
}
interface policyDocumentType {
  Version: string;
  Statement: policyDocumentStateType[];
}
interface policyDocumentStateType {
  Sid: string;
  Effect: string;
  Action: string[];
  Resource: string;
  Condition: ipAddressType;
}
interface ipAddressType {
  IpAddress: sourceIpType;
}
interface sourceIpType {
  "aws:SourceIp": string[];
}
export class IamUserStack extends cdk.Stack {
  private CIDR_REGEX = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(3[0-2]|[12]?[0-9])$/;
  constructor(scope: cdk.Construct, id: string, props: IamUserStacksProps) {
    super(scope, id, props);
    props.strictedCidrs.forEach((ip) => {
      if (!this.CIDR_REGEX.test(ip)) {
        throw new Error("invalid ips");
      }
    });
    let prefix = "";
    if (
      props.prefix &&
      props.prefix.length > 0 &&
      !props.prefix.match(/(master|prod|production|main)/)
    ) {
      prefix = `${props.prefix}-`;
    }
    const group: iam.Group = new iam.Group(
      this,
      `${prefix}${props.groupName}`,
      {
        groupName: `${prefix}${props.groupName}`,
      }
    );
    const baseGroupInlinePolicy: iam.Policy = new iam.Policy(
      this,
      group + "InlinePolicy",
      {
        document: iam.PolicyDocument.fromJson(
          this.generateDefaultPolicyJson(props.strictedCidrs)
        ),
      }
    );
    group.attachInlinePolicy(baseGroupInlinePolicy);
    const users = props.userNames;
    users.forEach((u) => {
      const userName = `${prefix}${u}`;
      const user: iam.User = this.createIamUser(userName, group);
      const inlinePolicy: iam.Policy = new iam.Policy(
        this,
        userName + "-policy",
        {
          document: iam.PolicyDocument.fromJson(
            this.generateInlinePolicyJson(props.strictedCidrs, userName)
          ),
        }
      );
      user.attachInlinePolicy(inlinePolicy);
    });
  }
  createIamUser(userName: string, group: iam.Group): iam.User {
    const user: iam.User = new iam.User(this, userName, {
      userName: userName,
    });
    user.addToGroup(group);
    return user;
  }
  generateInlinePolicyJson(
    ips: string[],
    username: string
  ): policyDocumentType {
    const userPolicyDocument = {
      Version: "2008-10-17",
      Statement: [
        {
          Sid: "AllowPush",
          Effect: "Allow",
          Action: [
            "ecr:CompleteLayerUpload",
            "ecr:InitiateLayerUpload",
            "ecr:PutImage",
            "ecr:UploadLayerPart",
            "ecr-public:CompleteLayerUpload",
            "ecr-public:InitiateLayerUpload",
            "ecr-public:PutImage",
            "ecr-public:UploadLayerPart",
          ],
          Resource: `arn:aws:ecr:*:*:repository/${username}/*`,
          Condition: {
            IpAddress: {
              "aws:SourceIp": ips,
            },
          },
        },
      ],
    };
    return userPolicyDocument;
  }
  generateDefaultPolicyJson(ips: string[]): policyDocumentType {
    const policyDocument = {
      Version: "2012-10-17",
      Statement: [
        {
          Sid: "VisualEditor0",
          Effect: "Allow",
          Action: [
            "ecr:DescribeImageScanFindings",
            "ecr:GetLifecyclePolicyPreview",
            "ecr:GetDownloadUrlForLayer",
            "ecr:BatchGetImage",
            "ecr:DescribeImages",
            "ecr:GetAuthorizationToken",
            "ecr:DescribeRepositories",
            "ecr:ListTagsForResource",
            "ecr:ListImages",
            "ecr:BatchCheckLayerAvailability",
            "ecr:GetRepositoryPolicy",
            "ecr:GetLifecyclePolicy",
            "ecr-public:GetAuthorizationToken",
            "sts:GetServiceBearerToken",
            "ecr-public:BatchCheckLayerAvailability",
            "ecr-public:GetRepositoryPolicy",
            "ecr-public:DescribeRepositories",
            "ecr-public:DescribeRegistries",
            "ecr-public:DescribeImages",
            "ecr-public:DescribeImageTags",
            "ecr-public:GetRepositoryCatalogData",
            "ecr-public:GetRegistryCatalogData",
          ],
          Resource: "*",
          Condition: {
            IpAddress: {
              "aws:SourceIp": ips,
            },
          },
        },
      ],
    };
    return policyDocument;
  }
}
