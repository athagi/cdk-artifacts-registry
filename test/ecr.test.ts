import {
  expect as expectCDK,
  haveResource,
  countResources,
  SynthUtils,
} from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import * as Ecr from "../lib/ecr-stack";

describe("fine grained tests", () => {
  it("ECR projects Created", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "test";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    repoNames.forEach((repo) => {
      expectCDK(stack).to(
        haveResource("AWS::ECR::Repository", {
          RepositoryName: `${prefix}-${repo}`,
        })
      );
    });
  });

  it("ECR projects Created without prefix", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
    });
    // THEN
    repoNames.forEach((repo) => {
      expectCDK(stack).to(
        haveResource("AWS::ECR::Repository", {
          RepositoryName: `${repo}`,
        })
      );
    });
  });

  it("ECR projects Created with master", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "master";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    repoNames.forEach((repo) => {
      expectCDK(stack).to(
        haveResource("AWS::ECR::Repository", {
          RepositoryName: `${repo}`,
        })
      );
    });
  });

  it("ECR projects Created with main", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "main";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    repoNames.forEach((repo) => {
      expectCDK(stack).to(
        haveResource("AWS::ECR::Repository", {
          RepositoryName: `${repo}`,
        })
      );
    });
  });

  it("ECR repository count", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    expectCDK(stack).to(
      countResources("AWS::ECR::Repository", repoNames.length)
    );
  });
});

describe("snapshot test", () => {
  it("repository is created", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "develop";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("repository is created", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const prefix = "";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});

describe("Validation tests", () => {
  // todo うまい書き方がわからない
  it("Validate ecr lifecycle rule", () => {
    const app = new cdk.App();
    // WHEN
    const repoNames = ["testrepo1", "testrepo2", "testrepo3", "testrepo4"];
    const lifecyclePolicyText = JSON.stringify({
      rules: [
        {
          rulePriority: 1,
          selection: {
            tagStatus: "any",
            countType: "imageCountMoreThan",
            countNumber: 10,
          },
          action: {
            type: "expire",
          },
        },
      ],
    });
    const prefix = "";
    const stack = new Ecr.RepositoriesStack(app, "MyTestStack", {
      repoNames: repoNames,
      lifecycleRule: Ecr.RepositoriesStack.LIFECYCLERULE,
      prefix: prefix,
    });
    // THEN
    repoNames.forEach((repo) => {
      expectCDK(stack).to(
        haveResource("AWS::ECR::Repository", {
          RepositoryName: repo,
          ImageScanningConfiguration: {
            scanOnPush: true,
          },
          LifecyclePolicy: {
            LifecyclePolicyText: lifecyclePolicyText,
          },
        })
      );
    });
  });
});
