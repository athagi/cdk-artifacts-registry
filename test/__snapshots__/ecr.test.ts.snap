// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test repository is created 1`] = `
Object {
  "Resources": Object {
    "developtestrepo1152856C2": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "develop-testrepo1",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "developtestrepo272F0705A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "develop-testrepo2",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "developtestrepo3DC9F216C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "develop-testrepo3",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "developtestrepo46EE393FB": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "develop-testrepo4",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`snapshot test repository is created 2`] = `
Object {
  "Resources": Object {
    "testrepo1698DDA31": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "testrepo1",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "testrepo27F5B92FC": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "testrepo2",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "testrepo3D48A6826": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "testrepo3",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "testrepo4DD3C7AD7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "scanOnPush": true,
        },
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "testrepo4",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
